[Bug] [connector][hive] java.net.SocketTimeoutException: Read timed out

### Search before asking

- [X] I had searched in the [issues](https://github.com/DTStack/chunjun/issues) and found no similar issues.


### What happened

当使用hive jdbc 读取hive时，偶发读取超时报错


```

2023-05-22 09:23:16,159 INFO  org.apache.flink.runtime.dispatcher.MiniDispatcher           [] - Job 80f33e671476505a82e264be2e13618a reached terminal state FAILED.
org.apache.flink.runtime.JobException: Recovery is suppressed by NoRestartBackoffTimeStrategy
	at org.apache.flink.runtime.executiongraph.failover.flip1.ExecutionFailureHandler.handleFailure(ExecutionFailureHandler.java:139)
	at org.apache.flink.runtime.executiongraph.failover.flip1.ExecutionFailureHandler.getFailureHandlingResult(ExecutionFailureHandler.java:83)
	at org.apache.flink.runtime.scheduler.DefaultScheduler.recordTaskFailure(DefaultScheduler.java:256)
	at org.apache.flink.runtime.scheduler.DefaultScheduler.handleTaskFailure(DefaultScheduler.java:247)
	at org.apache.flink.runtime.scheduler.DefaultScheduler.onTaskFailed(DefaultScheduler.java:240)
	at org.apache.flink.runtime.scheduler.SchedulerBase.onTaskExecutionStateUpdate(SchedulerBase.java:738)
	at org.apache.flink.runtime.scheduler.SchedulerBase.updateTaskExecutionState(SchedulerBase.java:715)
	at org.apache.flink.runtime.scheduler.SchedulerNG.updateTaskExecutionState(SchedulerNG.java:78)
	at org.apache.flink.runtime.jobmaster.JobMaster.updateTaskExecutionState(JobMaster.java:477)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.lambda$handleRpcInvocation$1(AkkaRpcActor.java:309)
	at org.apache.flink.runtime.concurrent.akka.ClassLoadingUtils.runWithContextClassLoader(ClassLoadingUtils.java:83)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcInvocation(AkkaRpcActor.java:307)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleRpcMessage(AkkaRpcActor.java:222)
	at org.apache.flink.runtime.rpc.akka.FencedAkkaRpcActor.handleRpcMessage(FencedAkkaRpcActor.java:84)
	at org.apache.flink.runtime.rpc.akka.AkkaRpcActor.handleMessage(AkkaRpcActor.java:168)
	at akka.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:24)
	at akka.japi.pf.UnitCaseStatement.apply(CaseStatements.scala:20)
	at scala.PartialFunction.applyOrElse(PartialFunction.scala:123)
	at scala.PartialFunction.applyOrElse$(PartialFunction.scala:122)
	at akka.japi.pf.UnitCaseStatement.applyOrElse(CaseStatements.scala:20)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:171)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:172)
	at akka.actor.Actor.aroundReceive(Actor.scala:537)
	at akka.actor.Actor.aroundReceive$(Actor.scala:535)
	at akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:220)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:580)
	at akka.actor.ActorCell.invoke(ActorCell.scala:548)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
	at akka.dispatch.Mailbox.run(Mailbox.scala:231)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: java.lang.IllegalArgumentException: open() failed.org.apache.thrift.transport.TTransportException: java.net.SocketTimeoutException: Read timed out
 querySQL: SELECT * FROM (xxxx) chunjun_tmp WHERE  1=1 
	at com.dtstack.chunjun.connector.hive.source.HiveInputFormat.openInternal(HiveInputFormat.java:108)
	at com.dtstack.chunjun.source.format.BaseRichInputFormat.open(BaseRichInputFormat.java:162)
	at com.dtstack.chunjun.source.DtInputFormatSourceFunction.run(DtInputFormatSourceFunction.java:118)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:110)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:67)
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:333)
Caused by: java.sql.SQLException: org.apache.thrift.transport.TTransportException: java.net.SocketTimeoutException: Read timed out
	at org.apache.hive.jdbc.HiveStatement.execute(HiveStatement.java:311)
	at org.apache.hive.jdbc.HiveStatement.executeQuery(HiveStatement.java:392)
	at com.dtstack.chunjun.connector.hive.source.HiveInputFormat.executeQuery(HiveInputFormat.java:128)
	at com.dtstack.chunjun.connector.hive.source.HiveInputFormat.openInternal(HiveInputFormat.java:92)
	... 5 more
Caused by: org.apache.thrift.transport.TTransportException: java.net.SocketTimeoutException: Read timed out
	at org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:129)
	at org.apache.thrift.transport.TTransport.readAll(TTransport.java:86)
	at org.apache.thrift.transport.TSaslTransport.readLength(TSaslTransport.java:376)
	at org.apache.thrift.transport.TSaslTransport.readFrame(TSaslTransport.java:453)
	at org.apache.thrift.transport.TSaslTransport.read(TSaslTransport.java:435)
	at org.apache.thrift.transport.TSaslClientTransport.read(TSaslClientTransport.java:37)
	at org.apache.thrift.transport.TTransport.readAll(TTransport.java:86)
	at org.apache.thrift.protocol.TBinaryProtocol.readAll(TBinaryProtocol.java:429)
	at org.apache.thrift.protocol.TBinaryProtocol.readI32(TBinaryProtocol.java:318)
	at org.apache.thrift.protocol.TBinaryProtocol.readMessageBegin(TBinaryProtocol.java:219)
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:77)
	at org.apache.hive.service.cli.thrift.TCLIService$Client.recv_GetOperationStatus(TCLIService.java:409)
	at org.apache.hive.service.cli.thrift.TCLIService$Client.GetOperationStatus(TCLIService.java:396)
	at org.apache.hive.jdbc.HiveStatement.execute(HiveStatement.java:280)
	... 8 more
Caused by: java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.socketRead(SocketInputStream.java:116)
	at java.net.SocketInputStream.read(SocketInputStream.java:170)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at java.io.BufferedInputStream.fill(BufferedInputStream.java:246)
	at java.io.BufferedInputStream.read1(BufferedInputStream.java:286)
	at java.io.BufferedInputStream.read(BufferedInputStream.java:345)
	at org.apache.thrift.transport.TIOStreamTransport.read(TIOStreamTransport.java:127)
	... 21 more
2023-05-22 09:23:16,168 INFO  org.apache.flink.runtime.dispatcher.MiniDispatcher           [] - Job 80f33e671476505a82e264be2e13618a has been registered for cleanup in the JobResultStore after reaching a terminal state.

```



### What you expected to happen

--

### How to reproduce



```
CREATE TABLE source
(
  `tday` string,
  `name` string,
  `step_one_num` bigint,
  `step_two_num` bigint,
  `step_three_num` bigint
) WITH (
      'file-type' = '',
      'password' = 'xxx',
      'connector' = 'hive-x',
      'properties.mapred.job.name' = 'xxx',
      'table-name' = '',
      'default-fs' = '',
      'scan.custom-sql' = 'select xxx from xxx ',
      'url' = 'jdbc:hive2://xxx:10000/xxx;connectTimeout=3000',
      'username' = 'xxx'
      );
CREATE TABLE sink
(
  `tday` varchar,
  `name` varchar,
  `step_one_num` bigint,
  `step_two_num` bigint,
  `step_three_num` bigint
) WITH (
      'password' = 'xxx',
      'connector' = 'mysql-x',
      'sink.buffer-flush.interval' = '2000',
      'sink.post-sql' = '',
      'sink.buffer-flush.max-rows' = '1024',
      'sink.all-replace' = 'false',
      'sink.pre-sql' = '',
      'table-name' = 'd',
      'sink.parallelism' = '1',
      'url' = 'xxxx',
      'username' = 'xxx'
      );
insert into sink 
select 
CAST(`tday` AS varchar) as `tday`,
CAST(`name` AS varchar) as `name`,
`step_one_num` as `step_one_num`,
`step_two_num` as `step_two_num`,
`step_three_num` as `step_three_num`
 from source;

```


### Anything else

_No response_

### Version

master

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/chunjun-connectors/chunjun-connector-hive/src/main/java/com/dtstack/chunjun/connector/hive/source/HiveInputFormat.java
+++ b/chunjun-connectors/chunjun-connector-hive/src/main/java/com/dtstack/chunjun/connector/hive/source/HiveInputFormat.java
@@ -49,6 +49,9 @@ public class HiveInputFormat extends JdbcInputFormat {
         connectionInfo.setJdbcUrl(jdbcConfig.getJdbcUrl());
         connectionInfo.setUsername(jdbcConfig.getUsername());
         connectionInfo.setPassword(jdbcConfig.getPassword());
+        if (jdbcConfig.getQueryTimeOut() > 0) {
+            connectionInfo.setTimeout(jdbcConfig.getQueryTimeOut());
+        }
         this.currentJdbcInputSplit = (JdbcInputSplit) inputSplit;
         initMetric(currentJdbcInputSplit);
         if (!canReadData(currentJdbcInputSplit)) {

--- a/chunjun-connectors/chunjun-connector-hive/src/main/java/com/dtstack/chunjun/connector/hive/table/HiveDynamicTableFactory.java
+++ b/chunjun-connectors/chunjun-connector-hive/src/main/java/com/dtstack/chunjun/connector/hive/table/HiveDynamicTableFactory.java
@@ -190,6 +190,7 @@ public class HiveDynamicTableFactory implements DynamicTableSinkFactory, Dynamic
         options.add(HiveOptions.PASSWORD);
         options.add(HiveOptions.PARTITION_TYPE);
         options.add(HiveOptions.PARTITION);
+        options.add(SCAN_QUERY_TIMEOUT);
 
         options.add(SCAN_CUSTOM_SQL);
 

--- a/chunjun-connectors/chunjun-connector-hive/src/main/java/com/dtstack/chunjun/connector/hive/util/HiveDbUtil.java
+++ b/chunjun-connectors/chunjun-connector-hive/src/main/java/com/dtstack/chunjun/connector/hive/util/HiveDbUtil.java
@@ -175,7 +175,7 @@ public class HiveDbUtil {
             ClassUtil.forName(
                     "org.apache.hive.jdbc.HiveDriver",
                     Thread.currentThread().getContextClassLoader());
-            return getHiveConnection(connectionInfo.getJdbcUrl(), prop);
+            return getHiveConnection(connectionInfo, prop);
         } catch (SQLException e) {
             if (SQLSTATE_USERNAME_PWD_ERROR.equals(e.getSQLState())) {
                 throw new RuntimeException("user name or password wrong.");
@@ -197,7 +197,9 @@ public class HiveDbUtil {
         }
     }
 
-    private static Connection getHiveConnection(String url, Properties prop) throws Exception {
+    private static Connection getHiveConnection(ConnectionInfo connectionInfo, Properties prop)
+            throws Exception {
+        String url = connectionInfo.getJdbcUrl();
         Matcher matcher = HIVE_JDBC_PATTERN.matcher(url);
         String db = null;
         String host = null;
@@ -213,6 +215,7 @@ public class HiveDbUtil {
         if (StringUtils.isNotEmpty(host) && StringUtils.isNotEmpty(db)) {
             param = param == null ? "" : param;
             url = String.format("jdbc:hive2://%s:%s/%s", host, port, param);
+            DriverManager.setLoginTimeout(connectionInfo.getTimeout());
             Connection connection = DriverManager.getConnection(url, prop);
             if (StringUtils.isNotEmpty(db)) {
                 try (Statement statement = connection.createStatement()) {

